### PR TITLE
add-WebSessionFortinetFortigate

### PR DESCRIPTION
**Pull request by KustoKing**
Required items, please complete
   
   Change(s):
   - Add Fortinet Fortigate Websession parser

   Reason for Change(s):
   - New functionality

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

**Comment by vakohl** [comment]
@KustoKing can you please attach sample logs to review the parser?

File: Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+      , EventOriginalSeverity = LogSeverity
+      , EventProduct = DeviceProduct
+      , EventProductVersion = DeviceVersion
+      , SrcBytes = SentBytes
+      , SrcInterfaceName = DeviceInboundInterface
+      , SrcIpAddr = SourceIP
+      , SrcPortNumber = SourcePort
+      , DvcId = DeviceExternalID
+      , EventUid = _ItemId
+    | invoke _ASIM_ResolveNetworkProtocol ('Protocol')
```
Comment (vakohl) [inline Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml:136]: use of this?

File: Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+"(2) Info: Missing optional field [HttpRequestHeaderCount]"
+"(2) Info: Missing optional field [HttpRequestMethod]"
+"(2) Info: Missing optional field [HttpRequestTime]"
+"(2) Info: Missing optional field [HttpRequestXff]"
+"(2) Info: Missing optional field [HttpResponseBodyBytes]"
+"(2) Info: Missing optional field [HttpResponseCacheControl]"
+"(2) Info: Missing optional field [HttpResponseExpires]"
+"(2) Info: Missing optional field [HttpResponseHeaderCount]"
+"(2) Info: Missing optional field [HttpResponseTime]"
+"(2) Info: Missing optional field [HttpUserAgent]"
```
Comment (vakohl) [inline Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv:75]: we are not getting this field from Fortinet?

File: Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+"(1) Warning: Missing recommended field [DstHostname]"
+"(1) Warning: Missing recommended field [DvcDomain]"
+"(1) Warning: Missing recommended field [DvcHostname]"
+"(1) Warning: Missing recommended field [DvcIpAddr]"
+"(1) Warning: Missing recommended field [SrcDomain]"
+"(1) Warning: Missing recommended field [SrcHostname]"
+"(2) Info: Missing optional alias [InnerVlanId] aliasing non-existent column [SrcVlanId]"
+"(2) Info: Missing optional alias [OuterVlanId] aliasing non-existent column [DstVlanId]"
+"(2) Info: Missing optional alias [UserAgent] aliasing non-existent column [HttpUserAgent]"
+"(2) Info: Missing optional alias [User] aliasing non-existent column [SrcUsername]"
```
Comment (vakohl) [inline Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv:13]: not getting any user info too?

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml [hunk 4]
```diff
@@ -0,0 +1,126 @@
+Parser:
+  Title: Web Session ASIM parser for Fortinet FortiGate
```
Comment (vakohl) [inline Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:2]: please attach sample logs to test parser

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+- Title: web log fields
+  Link: https://docs.fortinet.com/document/fortigate/7.4.0/fortios-log-message-reference/400992  
+- Title: Fortinet FortiGate CEF setup
+  Link: https://community.fortinet.com/t5/FortiGate/Technical-Tip-Integrate-FortiGate-with-Microsoft-Sentinel/ta-p/199709
+- Title: Fortinet FortiGate syslogd documentation
+  Link: https://docs.fortinet.com/document/fortigate/7.2.0/cli-reference/450620/config-log-syslogd-setting
+Description: |
+  This ASIM parser supports filtering and normalizing Fortinet FortiGate logs produced by the Microsoft Sentinel Fortinet connector to the ASIM Web Session normalized schema.
+ParserName: ASimWebSessionFortinetFortiGate
+EquivalentBuiltInParser: _ASim_NetworkSession_FortinetFortiGate
```
Comment (vakohl) [inline Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:24]: _ASim_WebSession_FortinetFortiGate?

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml (hunk 4 above)
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:2]: Solved as discussed

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    , "blocked","Deny","Failure"
+  ];
+  // -- See https://docs.fortinet.com/document/fortigate/7.2.4/fortios-log-message-reference/671442/cef-priority-levels
+  let SeverityLookup = datatable (EventOriginalSeverity:string, EventSeverity:string)
+  [
+    "1", "Informational", // Debug
+    "2", "Informational", // Information
+    "3", "Informational", // Notification
+    "4", "Low", // Warning
+    "5", "Low", // Error
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:42]: Yeah I will swap it arround. Also need to fix it at Fortinet network session

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    "6", "Critical", // High
+    "7", "Alert", // Medium
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:44]: Sorted

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    "8", "High" // Emergency
+  ];
+  let Parser=(disabled:bool=false){
+    CommonSecurityLog
+    | where not(disabled)
+    | where DeviceVendor == "Fortinet" 
+        and DeviceProduct startswith "FortiGate"
+        and (Activity == "webfilter utm passthrough"
+            or Activity == "webfilter utm blocked")
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:53]: We can add it, but fortinet only returns lowercase as far as I see.

File: Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+  ){
+    let src_or_any                = set_union(
+      srcipaddr_has_any_prefix,
+      ipaddr_has_any_prefix
+    ); 
+    let remove_protocol_from_list = (list:dynamic)
+    {
+        print list
+        | mv-apply l = print_0 to typeof(string) on
+          ( extend l = replace_regex (tostring(l), "^(?i:.*?)://", "") )
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml:73]: Yes, thanks

File: Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        temp_SrcMatch , "SrcIpAddr",
+        temp_DstMatch, "DstIpAddr",
+        "No match") 
+    | where ASimMatchingIpAddr != "No match" 
+    | project-away temp_*
+    | extend EventResultDetails = "NA"
+    | where (array_length(eventresultdetails_in) == 0 or tostring(EventResultDetails) has_any(eventresultdetails_in)) 
+    | lookup EventLookup on DeviceAction 
+    | where (eventresult == '*' or EventResult =~ eventresult)
+    | parse AdditionalExtensions with * "referralurl=" referralurl:string ";" *
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml:116]: Yes

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml (hunk 5 above)
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:24]: Yes

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    | parse AdditionalExtensions with * "referralurl=" referralurl:string ";" *
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:54]: Yes

File: Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    // -- See https://docs.fortinet.com/document/fortigate/7.2.4/fortios-log-message-reference/671442/cef-priority-levels
+    let SeverityLookup = datatable (EventOriginalSeverity:string, EventSeverity:string)
+    [
+        "1", "Informational", // Debug
+        "2", "Informational", // Information
+        "3", "Informational", // Notification
+        "4", "Low", // Warning
+        "5", "Low", // Error
+        "6", "Critical", // High
+        "7", "Alert", // Medium
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml:90]: Fixed

File: Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+    ];     
+    CommonSecurityLog
+    | where not(disabled)
+    | where (isnull(starttime) or TimeGenerated>=starttime) and (isnull(endtime) or TimeGenerated<=endtime)
+    | where DeviceVendor == "Fortinet" 
+      and DeviceProduct startswith "FortiGate"
+      and (Activity == "webfilter utm passthrough"
+          or Activity == "webfilter utm blocked")
+    | where (array_length(url_has_any) == 0 or RequestURL has_any (remove_protocol_from_list(url_has_any)))
+    | where array_length(httpuseragent_has_any) == 0
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml:101]: This is in extended logging, added this to parser and reference to information

File: Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml [hunk 14]
```diff
@@ -0,0 +1,186 @@
+Parser:
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml:1]: Yes

File: Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+        , "blocked","Deny","Failure"
+    ];
+    // -- See https://docs.fortinet.com/document/fortigate/7.2.4/fortios-log-message-reference/671442/cef-priority-levels
+    let SeverityLookup = datatable (EventOriginalSeverity:string, EventSeverity:string)
+    [
+        "1", "Informational", // Debug
+        "2", "Informational", // Information
+        "3", "Informational", // Notification
+        "4", "Low", // Warning
+        "5", "Low", // Error
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml:88]: When we place Error and Alert in context I think Error is low. Or we should elevate Alert to High.

https://docs.fortinet.com/document/fortigate/7.2.4/fortios-log-message-reference/671442/cef-priority-levels
Error : An error exists and functionality could be affected.
Alert : Immediate action required. Used in security logs.

File: Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv (hunk 2 above)
Comment (KustoKing) [reply Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv:75]: Sorted

File: Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv (hunk 3 above)
Comment (KustoKing) [reply Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv:13]: Did not find user info

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+      , SrcInterfaceName = DeviceInboundInterface
+      , SrcIpAddr = SourceIP
+      , SrcPortNumber = SourcePort
+      , DvcId = DeviceExternalID
+      , EventUid = _ItemId
+    | invoke _ASIM_ResolveNetworkProtocol ('Protocol')
+    | project-rename DvcOriginalAction = DeviceAction
+    | parse-kv AdditionalExtensions as (
+            httpmethod:string,
+            agent:string
```
Comment (vakohl) [inline Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:81]: Found different logs. Have shared an email with you for updating logic of these two fields. Please have a look.

**Review by vakohl — CHANGES_REQUESTED** [review]

**Comment by v-atulyadav** [comment]
Hi @KustoKing, please check comments from @vakohl and act accordingly. Thanks

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml (hunk 16 above)
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:81]: Switched logic

File: Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv [hunk 17]
```diff
@@ -0,0 +1,111 @@
+﻿Result
+"(0) Error: Missing recommended alias [HttpStatusCode] aliasing existing column [EventResultDetails]"
+"(1) Warning: Missing recommended field [ASimMatchingIpAddr]"
+"(1) Warning: Missing recommended field [DstHostname]"
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv:4]: Check

File: Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+"(1) Warning: Missing recommended field [DvcDomain]"
+"(1) Warning: Missing recommended field [DvcHostname]"
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv:6]: Done

File: Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+      , EventProduct          = DeviceProduct
+      , EventProductVersion   = DeviceVersion
+      , SrcBytes              = SentBytes
+      , SrcInterfaceName      = DeviceInboundInterface
+      , SrcIpAddr             = SourceIP
+      , SrcPortNumber         = SourcePort
+      , DvcId                 = DeviceExternalID
+      , EventUid              = _ItemId
+      , DstHostname           = DestinationHostName
+      , SrcHostname           = SourceHostname
```
Comment (vakohl) [inline Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml:139]: this is giving error. should be
, SrcHostname           = SourceHostName

File: Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv [hunk 20]
```diff
@@ -0,0 +1,111 @@
+﻿Result
+"(0) Error: Missing recommended alias [HttpStatusCode] aliasing existing column [EventResultDetails]"
+"(1) Warning: Missing recommended field [ASimMatchingIpAddr]"
```
Comment (vakohl) [inline Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv:3]: pls project this: ASimMatchingIpAddr

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+      , EventProduct          = DeviceProduct
+      , EventProductVersion   = DeviceVersion
+      , SrcBytes              = SentBytes
+      , SrcInterfaceName      = DeviceInboundInterface
+      , SrcIpAddr             = SourceIP
+      , SrcPortNumber         = SourcePort
+      , DvcId                 = DeviceExternalID
+      , EventUid              = _ItemId
+      , DstHostname           = DestinationHostName
+      , SrcHostname           = SourceHostname
```
Comment (vakohl) [inline Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:79]: SourceHostname should be SourceHostName

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml (hunk 19 above)
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/vimWebSessionFortinetFortiGate.yaml:139]: Fixed, thanks

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml (hunk 21 above)
Comment (KustoKing) [reply Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:79]: Fixed

File: Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv (hunk 20 above)
Comment (KustoKing) [reply Parsers/ASimWebSession/test/Fortinet_FortiGate_WebSession_SchemaTest.csv:3]: Sorted

**Review by KustoKing — COMMENTED** [review]
Changes should be there

**Comment by v-atulyadav** [comment]
Hi @KustoKing, 
Thank you for these fixes. We will provide an update as soon as possible.

**Review by vakohl — DISMISSED** [review]

**Comment by v-atulyadav** [comment]
Hi @KustoKing, 
The contents have been reviewed and approved, please resolve branch conflicts so that we can proceed with merging.
Thanks

**Review by vakohl — APPROVED** [review]

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**